File: c20_palindrome_string/L131-m.cpp
//
// Created by 李恒 on 2020/7/1.
//
#include <cstring>
#include <iostream>
#include <vector>

using  namespace std;
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        int sz = s.size();
        vector<vector<bool>> dp(sz, vector<bool>(sz, false));
        for(int i=0; i<sz; ++i){
            for(int j=i; j>=0; j--){
                if(s[i]==s[j] && ((i-j)<2 || dp[j+1][i-1]))
                    dp[j][i] = true; //注意是dp[j][i] 不是dp[i][j]
            }
        }
        vector<string> tmp;
        dfs(res, dp, 0, s, tmp);
        return res;

    }

    void dfs(vector<vector<string>>& res, vector<vector<bool>> dp, int start, string s, vector<string>& tmp  ){
        if(start == s.size()) res.push_back(tmp);
        for(int j=start; j<s.size(); ++j){
            if(dp[start][j]){
                tmp.push_back(s.substr(start, j-start+1));
                dfs(res, dp, j+1, s, tmp);
                tmp.pop_back();
            }
        }
    }

};

File: python-algorithm/ds_6_string/palindrome_string/L131-m.py
# -*- coding: utf-8 -*-

# 131 分割回文串
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        sz = len(s)
        dp = [[0 for i in range(sz)] for i in range(sz)]
        # 即每次遍历以当前元素作为回文串最后一个值
        for i in range(sz):
            j = i
            while j>=0:
                if s[j]==s[i] and (i-j<2 or dp[j+1][i-1]):
                    dp[j][i] = 1
                j -= 1
        res = []
        tmp = []
        self.dfs(s, 0, dp, tmp, res)
        return res

    def dfs(self, s, start, dp, tmp, res):
        if start==len(s):
            res.append(tmp[:])
        for j in range(start, len(s)):
            if dp[start][j]:
                tmp.append(s[start:j+1])
                self.dfs(s, j+1, dp, tmp, res)
                tmp.pop()